File: typescript/src/pages/heroku/frames.page.ts
import { BasePage } from '../base.page';

/** Frames hub (/frames) with nested-frames and iframe-editor support. */
export class FramesPage extends BasePage {
  async openNestedFrames(): Promise<void> {
    await this.page.getByRole('link', { name: 'Nested Frames', exact: true }).click();
  }

  async openIframeEditor(): Promise<void> {
    await this.page.getByRole('link', { name: 'iFrame', exact: true }).click();
  }

  /** Body text of a nested frame: 'left' | 'middle' | 'right' | 'bottom'. */
  async nestedFrameText(position: 'left' | 'middle' | 'right' | 'bottom'): Promise<string> {
    const text =
      position === 'bottom'
        ? await this.page.frameLocator("frame[src='/frame_bottom']").locator('body').innerText()
        : await this.page
            .frameLocator("frame[src='/frame_top']")
            .frameLocator(`frame[src='/frame_${position}']`)
            .locator('body')
            .innerText();
    return (text ?? '').trim();
  }

  /**
   * Assert the TinyMCE iframe is present and visible.
   * NOTE: the demo site's TinyMCE cloud key is out of editor loads, so the
   * editor renders read-only with no content — presence is what's asserted.
   */
  async assertEditorPresent(): Promise<void> {
    await this.page.locator('#mce_0_ifr').waitFor({ state: 'visible' });
  }
}

File: src/main/java/io/github/mb1lal/playwright/ui/pages/heroku/FramesPage.java
package io.github.mb1lal.playwright.ui.pages.herokuapp;

import io.github.mb1lal.playwright.base.TestContext;
import io.github.mb1lal.playwright.ui.pages.BasePage;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.AriaRole;
import lombok.extern.slf4j.Slf4j;

/**
 * Frames hub ({@code /frames}) with nested-frames and iframe-editor support.
 */
@Slf4j
public class FramesPage extends BasePage {

    public FramesPage(Page page, TestContext testContext) {
        super(page, testContext);
    }

    public void openNestedFrames() {
        click(page.getByRole(AriaRole.LINK,
                        new Page.GetByRoleOptions().setName("Nested Frames").setExact(true)),
                "Nested Frames link");
        page.waitForLoadState();
    }

    public void openIframeEditor() {
        click(page.getByRole(AriaRole.LINK,
                        new Page.GetByRoleOptions().setName("iFrame").setExact(true)),
                "iFrame link");
        page.waitForLoadState();
    }

    /** Body text of a nested frame: {@code left}, {@code middle}, {@code right} or {@code bottom}. */
    public String nestedFrameText(String position) {
        String text;
        if ("bottom".equalsIgnoreCase(position)) {
            text = page.frameLocator("frame[src='/frame_bottom']").locator("body").innerText();
        } else {
            text = page.frameLocator("frame[src='/frame_top']")
                    .frameLocator("frame[src='/frame_" + position.toLowerCase() + "']")
                    .locator("body").innerText();
        }
        log.debug("Nested frame '{}' text: '{}'", position, text.trim());
        return text.trim();
    }

    /**
     * Assert the TinyMCE iframe is present and visible.
     *
     * <p>NOTE: the demo site's TinyMCE cloud key is out of editor loads, so the
     * editor renders read-only with no content — reading or writing its body is
     * impossible. Presence of the iframe is what's asserted; this still proves
     * iframe handling, which is the point of the spec.
     */
    public void assertEditorPresent() {
        assertIsVisible(page.locator("#mce_0_ifr"), "TinyMCE editor iframe");
    }
}
